ssm/lds.py

```diff
@@ -578,29 +580,33 @@ def expected_log_joint(x):
 
             # Compute the gradient of the expected log joint at a point x
             grad_expected_log_joint = grad(expected_log_joint)
-            hess_expected_log_joint = None
 
-            # Compute Hessian
-            # hessian(expected_log_joint) # -> TD x TD dense matrix. BAD!
-            # d^2/dx^2 log p(yt | xt)  -> DxD for t=1:T
-            # d^2/dxt^2 log p(xt+1 | xt) -> DxD for t=1:T-1
-            # d^2/dxt+1^2 log p(xt+1 | xt) -> DxD for t=2:T
-            # d^2/dxt dxt+1 log p(xt+1 | xt) -> DxD for t=1:T-1
+            # Compute the expected Hessian, represented in blocks
+            def hessian_blocks(x):
 
-            def newtons_method(x0, grad, hessian_diag, hessian_lower_diag):
+                # The "mask" for x is all ones
+                x_mask = np.ones_like(x, dtype=bool)
+                hessian_diag, hessian_lower_diag = self.dynamics.hessian_expected_log_dynamics_prob(Ez, x, input, x_mask, tag)
+                hessian_diag += self.transitions.hessian_expected_log_trans_prob(x, input, x_mask, tag, Ezzp1)
+                #TODO include Ez
+                hessian_diag += self.emissions.hessian_log_emissions_prob(data, input, mask, tag, x)
+
+                return hessian_diag, hessian_lower_diag
+
+            def newtons_method(x0, grad, hessian_blocks):
                 # TODO: damping, etc
                 x = x0
                 while not is_converged:
-                    J_banded = blocks_to_bands(hessian_diag(x), hessian_lower_diag(x), lower=True)
+                    hessian_diag, hessian_lower_diag = hessian_blocks(x)
+                    J_banded = blocks_to_bands(hessian_diag, hessian_lower_diag, lower=True)
```

Comment:
I think theres a function in `ssm.primitives` for solving block tridiagonal systems.  It will handle converting to banded and calling scipy for you.